### PR TITLE
[flang][OpenMP] Try to unify induction var privatization for OMP regions.

### PR DESCRIPTION
**Pull request by ergawy**
This PR tries to unify the paths taken by flang to privatize the induction variable of loops. With the changes introduced here, both cases below are now privatized in the same way:

```fortran
subroutine foo
    implicit none
    integer :: i

    !$omp parallel default(private)
      do i = 1, 10
      end do
    !$omp end parallel
end subroutine
```

```fortran
subroutine foo
    implicit none
    integer :: i

    !$omp parallel private(i)
      do i = 1, 10
      end do
    !$omp end parallel
end subroutine
```

### CODE REVIEW

**Comment by ergawy** [comment]
With this change, both `private(i)` and `default(private)` go through `DataSharingProcessor::doPrivatize(...)` to emit the privatization logic for `i`.

If you agree with this change, we will be able to switch delayed privatization by default (see https://github.com/llvm/llvm-project/pull/90945). With the changes in this PR, we would get a crash for cases where `default(private)` is specified for the induction variable. The crash happened because `DataSharingProcessor::cloneSymbol(...)` would return directly for `OmpPreDetermined` symbols which prevents delayed privatization from being properly applied to the induction variable.

@reviewers Please take a look and let me know what you think.

**Comment by kiranchandramohan** [comment]
Could you check why the CI is failing?

>  Try to unify induction var privatization

Could you clarify that this is for sequential loops inside parallel region? We have a separate case for induction variables of worksharing loops. 


> If you agree with this change, we will be able to switch delayed privatization by default (see #90945). 

We should enable delayed privatization for all constructs that are currently supported before switching it on. Otherwise we will regress on existing enabled applications. This will involve adding translation support for constructs for which there is support already.
1. sections, single, worksharing loop, simd (?), task (only works for non-deferred execution).
2. In some cases a barrier has to be added after first-privatisation. I think this is the case where there are both first and lastprivate clauses specified.

Or do you mean selectiveley enabling privatization for the parallel construct?

**Comment by ergawy** [comment]
> Could you check why the CI is failing?

I added an assertion to test that pre-determined is not set. But seems to be causing issues. Looking into it now...

> Or do you mean selectiveley enabling privatization for the parallel construct?

Yes, that is what I meants since `parallel` is now supported across the pipeline so I was thinking of using delayed privatization for it as early as possible while working on other constructs.

**Comment by ergawy** [comment]
> Could you check why the CI is failing?

I added the assertion prematurely and it was triggering for IVs in worksharing loops as you (@kiranchandramohan) pointed out.

I will limit the scope of this PR for sequential loops inside `omp parallel` regions then.

File: flang/lib/Semantics/resolve-directives.cpp [hunk 1]
```diff
@@ -1674,7 +1674,6 @@ void OmpAttributeVisitor::ResolveSeqLoopIndexInParallelOrTaskConstruct(
   // parallel or task
   if (auto *symbol{ResolveOmp(iv, Symbol::Flag::OmpPrivate, targetIt->scope)}) {
     targetIt++;
-    symbol->set(Symbol::Flag::OmpPreDetermined);
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/resolve-directives.cpp:1677]: I would prefer retaining this. Can we skip if it is predetermined and a parallel region in the DataSharingprocessor?

**Comment by kiranchandramohan** [comment]
> > Or do you mean selectiveley enabling privatization for the parallel construct?
> 
> Yes, that is what I meants since `parallel` is now supported across the pipeline so I was thinking of using delayed privatization for it as early as possible while working on other constructs.

Makes sense.

**Comment by ergawy** [comment]
> For testing, please run the gfortran testsuite. Some of our CI will test this.
> 
> ```
> git clone https://github.com/llvm/llvm-test-suite.git
> cd llvm-test-suite
> mkdir build
> cd build
> cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=$HOME/llvm-project/build_release/bin/clang  -DCMAKE_CXX_COMPILER=$HOME/llvm-project/build_release/bin/clang++ -DCMAKE_Fortran_COMPILER=$HOME/llvm-project/build_release/bin/flang-new -DTEST_SUITE_FORTRAN=On -DTEST_SUITE_SUBDIRS=Fortran -DTEST_SUITE_FORTRAN_ISO_C_HEADER_DIR=$HOME/llvm-project/flang/include/flang ../
> make -j48
> NO_STOP_MESSAGE=1 $HOME/llvm-project/build_release/bin/llvm-lit -v  .
> ```
> 
> The fujitsu testsuite also is available. Some tests might be failing, so if you using this just ensure that there are no additional failures. https://github.com/fujitsu/compiler-test-suite

Thanks for the starting instructions.

Does this currently work for `flang-new` out of the box?
```
/home/kaergawy/git/llvm-test-suite/build/Fortran/UnitTests/hello && /home/kaergawy/git/trunk18.0/build/llvm-project-2/bin/flang-new -DNDEBUG  -w -Werror=date-time -c /home/kaergawy/git/llvm-test-suite/Fortran/UnitTests/hello/hello.f90 -o CMakeFiles/hello.dir/hello.f90.o
```

File: flang/lib/Semantics/resolve-directives.cpp (hunk 1 above)
Comment (ergawy) [reply flang/lib/Semantics/resolve-directives.cpp:1677]: Will do ....

**Comment by kiranchandramohan** [comment]
> > For testing, please run the gfortran testsuite. Some of our CI will test this.
> > ```
> > git clone https://github.com/llvm/llvm-test-suite.git
> > cd llvm-test-suite
> > mkdir build
> > cd build
> > cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=$HOME/llvm-project/build_release/bin/clang  -DCMAKE_CXX_COMPILER=$HOME/llvm-project/build_release/bin/clang++ -DCMAKE_Fortran_COMPILER=$HOME/llvm-project/build_release/bin/flang-new -DTEST_SUITE_FORTRAN=On -DTEST_SUITE_SUBDIRS=Fortran -DTEST_SUITE_FORTRAN_ISO_C_HEADER_DIR=$HOME/llvm-project/flang/include/flang ../
> > make -j48
> > NO_STOP_MESSAGE=1 $HOME/llvm-project/build_release/bin/llvm-lit -v  .
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The fujitsu testsuite also is available. Some tests might be failing, so if you using this just ensure that there are no additional failures. https://github.com/fujitsu/compiler-test-suite
> 
> Thanks for the starting instructions.
> 
> Does this currently work for `flang-new` out of the box?
> 
> ```
> /home/kaergawy/git/llvm-test-suite/build/Fortran/UnitTests/hello && /home/kaergawy/git/trunk18.0/build/llvm-project-2/bin/flang-new -DNDEBUG  -w -Werror=date-time -c /home/kaergawy/git/llvm-test-suite/Fortran/UnitTests/hello/hello.f90 -o CMakeFiles/hello.dir/hello.f90.o
> ```

The gfortran testsuite works out of the box. The Fujitsu testsuite has instructions for working with the llvm testsuite. https://github.com/fujitsu/compiler-test-suite/blob/main/RUN.md

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 2]
```diff
@@ -344,7 +353,6 @@ void DataSharingProcessor::collectSymbols(
     assert(curScope && "couldn't find current scope");
     if (isPrivatizable(*sym) && !symbolsInNestedRegions.contains(sym) &&
         !privatizedSymbols.contains(sym) &&
-        !sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined) &&
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: This test was preventing induction variables, when `default(private)` is specified, from being added to the list of symbols privatized by the `DataSharingProcessor`.

Comment (luporl) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: I've added this test to make `defaultSymbols` and `implicitSymbols` more accurate and to simplify `defaultPrivatize` and `implicitPrivatize`.

Also, implicit symbols (those whose DSA is implicitly determined) are only present when there are no default symbols in a given construct (see`collectImplicitSymbols`). If pre-determined symbols are included in `defaultSymbols` we can no longer know what is the set of default symbols.

Finally, in the non-delayed privatization path, pre-determined symbols were already being skipped:
```c++
void DataSharingProcessor::cloneSymbol(const Fortran::semantics::Symbol *sym) {
  // Privatization for symbols which are pre-determined (like loop index
  // variables) happen separately, for everything else privatize here.
  if (sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined))
    return;
```

As pre-determined symbols' DSAs don't depend on default clauses, I don't think they should be handled as default or implicit symbols. In your test program, if you remove `default(private)`, are induction variables correctly privatized?

Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: > Finally, in the non-delayed privatization path, pre-determined symbols were already being skipped:

Indeed. And this is what I am trying to prevent from happening here both for early and delayed privatization. AFAIU, I don't think there is a reason to have 2 different code-gen paths for pre-determined vs. non-pre-determined variables. Please let me know if I am mistaken here.

More context for this is in this Slack thread: https://flang-compiler.slack.com/archives/C01PY03PP9P/p1714729085154889.

>  if you remove default(private), are induction variables correctly privatized?

In that case, they are not privatized at all in the sense that their alloca's are emitted simply as part of emitting the code for the do concurrent for which they are used as induction variables. They are not marked as pre-determined.

Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: > more accurate and to simplify defaultPrivatize and implicitPrivatize.

Can you elaborate on what you mean by simplify?

Comment (luporl) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: > > Finally, in the non-delayed privatization path, pre-determined symbols were already being skipped:
> 
> Indeed. And this is what I am trying to prevent from happening here both for early and delayed privatization. AFAIU, I don't think there is a reason to have 2 different code-gen paths for pre-determined vs. non-pre-determined variables. Please let me know if I am mistaken here.
> 

I also don't think there is a reason to have 2 different paths to perform the privatization of pre-determined vs. non-predetermined variables. However, currently `DataSharingProcessor::collectSymbols` is intended to collect default and implicit symbols only. It won't even be called if there are no default or implicit symbols.

> More context for this is in this Slack thread: https://flang-compiler.slack.com/archives/C01PY03PP9P/p1714729085154889.
> 
> > if you remove default(private), are induction variables correctly privatized?
> 
> In that case, they are not privatized at all in the sense that their alloca's are emitted simply as part of emitting the code for the do concurrent for which they are used as induction variables. They are not marked as pre-determined (see [this](https://flang-compiler.slack.com/archives/C01PY03PP9P/p1714899066421309?thread_ts=1714729085.154889&cid=C01PY03PP9P)).

Thanks for pointing this out. I think this is the main problem. Pre-determined (private) variables must be privatized independently of the presence of `private` or `default` clauses. So there is something wrong with case 2, as the loop variable should also be marked as pre-determined in semantics phase.

But actually, it seems that currently loop variables are always being assumed to be pre-determined, which, AFAIU, is not always the case (see test/Semantics/OpenMP/symbol08.f90, for instance).

Right now it would probably be easier to refactor `genLoopVars`, to somehow pass the loop variables to the DataSharingProcessor, so that privatization happens only there.

Comment (luporl) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: > > more accurate and to simplify defaultPrivatize and implicitPrivatize.
> 
> Can you elaborate on what you mean by simplify?

Before #85989, `defaultSymbols` would have other unrelated symbols, that would be filtered at `defaultPrivatize` only. The filter part was moved to `collectSymbols`, which simplified `defaultPrivatize` and `implicitPrivatized` and allowed `symbolsInNestedRegions` to be removed.

Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:347]: Thanks for the clarification. I understand the context for that change better now.

> Right now it would probably be easier to refactor genLoopVars, to somehow pass the loop variables to the DataSharingProcessor, so that privatization happens only there.

I have a different proposal which I implemented in the latest commit on the PR. To keep the different lists of symbols (default, implicit, explicit) accurate, I created one more list of symbols to aggregate all symbols to be privatized including the pre-determined ones. And then, we would have a single `privatize` method to go over the aggregated list.

The problem with invoking the `DataSharingProcessor` from `genLoopVars` is that I think we will have to instantiate it from `genLoopVars`. I think instantiating and controlling instances of the DSP should be localized to the OMP constructs that need it in `Lower/OpenMP.cpp`.

Let me know if you disagree with the current approach.

**Review by luporl — COMMENTED** [review]
I like the idea of having `allPrivatizedSymbols` to let the privatized symbols be handled in the same way.

But the way it's implemented here, it won't work correctly for all cases.
`collectSymbols` will only collect:
1. Symbols whose DSA match that of a `default` clause
2. Symbols marked with OmpImplicit

So, as pre-determined symbols are never implicit, in a construct without a `default` or other explicit DSA clause, such as `private`, pre-determined symbols won't be collected for privatization.

IIUC, you want to collect and privatize all pre-determined symbols now, right?
If that's the case, you'll need to refactor `collectSymbols` and call it again, with flag = OmpPreDetermined.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 3]
```diff
@@ -57,7 +57,7 @@ void DataSharingProcessor::processStep2(mlir::Operation *op, bool isLoop) {
 }
 
 void DataSharingProcessor::insertDeallocs() {
-  for (const Fortran::semantics::Symbol *sym : privatizedSymbols)
+  for (const Fortran::semantics::Symbol *sym : allPrivatizedSymbols)
```
[the remark this answers is not in the document]
Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:61]: I think that was the case indeed.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 4]
```diff
@@ -289,12 +290,34 @@ void DataSharingProcessor::collectSymbolsInNestedRegions(
...
     if (nestedEval.hasNestedEvaluations()) {
       if (nestedEval.isConstruct())
-        // Recursively look for OpenMP constructs within `nestedEval`'s region
         collectSymbolsInNestedRegions(nestedEval, flag, symbolsInNestedRegions);
-      else
-        converter.collectSymbolSet(nestedEval, symbolsInNestedRegions, flag,
-                                   /*collectSymbols=*/true,
-                                   /*collectHostAssociatedSymbols=*/false);
+      else {
+        bool isOrderedConstruct = [&]() {
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:295]: @kiranchandramohan @luporl just to clarify this change:
Since `omp ordered` does not have its own data-sharing environment, I think we should consider symbols referenced by an `omp ordered` construct to be separate from the symbols referenced by the parent region.

The changes here make sure that for the following input, `i` to be collected as of the symbols to be privatized for the `omp do ordered` construct:
```fortran
  !$omp do ordered
  do i = 2, 10
    !$omp ordered
    a(i) = a(i-1) + 1
    !$omp end ordered
  end do
  !$omp end do
```

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 5]
```diff
@@ -203,17 +203,16 @@ static mlir::Operation *
...
 
   mlir::Type tempTy = converter.genType(*sym);
-  mlir::Value temp = firOpBuilder.create<fir::AllocaOp>(
-      loc, tempTy, /*pinned=*/true, /*lengthParams=*/mlir::ValueRange{},
-      /*shapeParams*/ mlir::ValueRange{},
-      llvm::ArrayRef<mlir::NamedAttribute>{
-          fir::getAdaptToByRefAttr(firOpBuilder)});
-  converter.bindSymbol(*sym, temp);
+
+  assert(converter.isPresentShallowLookup(*sym) &&
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:213]: Now that we extended the `DataSharingProcessor` to collect all kinds of private symbols, we can verify here that no allocation is required for `sym`.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 4 above)
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:295]: It is not clear to me why the ordered construct is a special case. Will the same issue be resolved by a fix similar to https://github.com/llvm/llvm-project/pull/90671 ?

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 6]
```diff
@@ -289,12 +290,34 @@ void DataSharingProcessor::collectSymbolsInNestedRegions(
        eval.getNestedEvaluations()) {
     if (nestedEval.hasNestedEvaluations()) {
       if (nestedEval.isConstruct())
-        // Recursively look for OpenMP constructs within `nestedEval`'s region
```
Comment (luporl) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:292]: I would leave this comment. It helps to understand what this part of the code is doing, that we only continue looking for OpenMP constructs recursively for non-OpenMP constructs. Otherwise `converter.collectSymbolSet` can be used to collect the remaining symbols.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 7]
```diff
@@ -330,25 +353,46 @@ void DataSharingProcessor::collectSymbols(
...
+
+    if (collectPreDetermined &&
+        sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined))
+      return true;
+
+    if (!sym->test(Fortran::semantics::Symbol::Flag::OmpImplicit) &&
+        !sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined))
+      return true;
+
+    return false;
```
Comment (luporl) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:385]:
```suggestion
    if (collectImplicit)
        return sym->test(Fortran::semantics::Symbol::Flag::OmpImplicit);
    if (collectPreDetermined)
        return sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined);
    return !sym->test(Fortran::semantics::Symbol::Flag::OmpImplicit) &&
        !sym->test(Fortran::semantics::Symbol::Flag::OmpPreDetermined);
```

`collectImplicit` and `collectPreDetermined` will never be both true and when collecting implicit or pre-determined symbols you'll only want to collect those symbols marked with the corresponding flag.
Also, no symbol should have both `OmpPreDetermined` and `OmpImplicit` flags.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 8]
```diff
@@ -365,6 +409,9 @@ void DataSharingProcessor::collectDefaultSymbols() {
                        defaultSymbols);
     }
   }
+
+  for (auto *sym : defaultSymbols)
+    allPrivatizedSymbols.insert(sym);
```
Comment (luporl) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:414]: This should not be needed, as default symbols are collected through `collectSymbols()`.

File: flang/lib/Lower/OpenMP/OpenMP.cpp (hunk 5 above)
Comment (luporl) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:213]: That's nice.

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 9]
```diff
@@ -1489,6 +1488,7 @@ genSimdOp(Fortran::lower::AbstractConverter &converter,
           Fortran::lower::pft::Evaluation &eval, mlir::Location loc,
           const List<Clause> &clauses) {
   fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+  symTable.pushScope();
```
Comment (luporl) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:1491]: Why is this needed now? To avoid name clashes, now that pre-determined variables are privatized by DataSharingProcessor?

File: flang/test/Lower/OpenMP/Todo/omp-default-clause-inner-loop.f90 [hunk 10]
```diff
@@ -8,7 +8,7 @@
 ! CHECK: omp.parallel   {
 ! EXPECTED: %[[PRIVATE_Y:.*]] = fir.alloca i32 {bindc_name = "y", pinned, uniq_name = "_QFEy"}
 ! EXPECTED: %[[PRIVATE_Z:.*]] = fir.alloca i32 {bindc_name = "z", pinned, uniq_name = "_QFEz"}
-! CHECK: %[[TEMP:.*]] = fir.alloca i32 {adapt.valuebyref, pinned}
+! CHECK: %[[TEMP:.*]] = fir.alloca i32 {bindc_name = "x"
```
Comment (luporl) [inline flang/test/Lower/OpenMP/Todo/omp-default-clause-inner-loop.f90:11]: Nit: if it's not too much work, it would be nice to keep braces balanced, using something like:
`{bindc_name = "x", {{.*}}}`

**Review by luporl — COMMENTED** [review]
In addition to @kiranchandramohan's comments, I have a few more, but these changes are looking good to me now, thanks for working on this.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 6 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:292]: Not sure why I removed. Thanks for noticing.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 7 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp]: Thanks for the suggestion. Done.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 8 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp]: Indeed. Done.

**Review by ergawy — COMMENTED** [review]
Thanks for the review. And sorry for the delay here, https://github.com/llvm/llvm-project/pull/90098 changed how things work and I had to find a solution for broken tests after merging with it.

File: flang/test/Lower/OpenMP/Todo/omp-default-clause-inner-loop.f90 (hunk 10 above)
Comment (ergawy) [reply flang/test/Lower/OpenMP/Todo/omp-default-clause-inner-loop.f90]: Done.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 4 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:296]: I can describe the problem in more details but cannot claim my solution is the proper one.

For the above example program, this is the output of `flang-new -fc1 -fopenmp -fdebug-unparse-with-symbols`:
```fortram
!DEF: /foo (Subroutine) Subprogram
subroutine foo
 !DEF: /foo/a ObjectEntity INTEGER(4)
 !DEF: /foo/i ObjectEntity INTEGER(4)
 integer a(10), i
!$omp do  ordered
 !DEF: /foo/OtherConstruct1/i (OmpPrivate, OmpPreDetermined) HostAssoc INTEGER(4)
 do i=2,10
!$omp ordered 
  !REF: /foo/a
  !REF: /foo/OtherConstruct1/i
  a(i) = a(i-1)+1
!$omp end ordered 
 end do
!$omp end do 
end subroutine
```

So, `i` is marked as being `REF`ed from within the `omp ordered` construct (not defined).

However, when you call `converter.collectSymbolSet(nestedEval, symbolsInNestedRegions, ...)` when `nestedEval` is the `omp ordered` construct, `i` will be one of the symbols collected in that nested region.

This in turn, causes `i` **not** to be privatized when privatizing symbols that should be privatized for the `omp do ordered` construct (`i` is skipped because it is referenced by a nested region).

Since, `omp ordered` does not have a data environment of its own (as far as I understand), I thought it would reasonable to not exclude the symbols collected for a nested eval when that nested eval is an `ordered` directive.

That said, I am not sure this is the proper location for that kind of fix. But a similar fix to what was introduced in https://github.com/llvm/llvm-project/pull/90671 does not seem to work. Tried adding that in `bool OmpAttributeVisitor::Pre(const parser::OpenMPBlockConstruct &x);` but I get the same problem described above.

File: flang/lib/Lower/OpenMP/OpenMP.cpp (hunk 9 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:1516]: Yes, in the `DataSharingProcessor` we `assert` whether a symbols with the same name was already in same scope that we use for the `eval` being processed.

If you have a combined construct like: `omp target simd`, each sub-construct will have its own data-sharing environment (a separate instance of the DSP). `i` will be `map_entries` element on the target, so it exists in the scope of the `target` sub-construct. However, we also need to privatize `i` for the `simd` construct and use the private copy. This is what was being done before already but we just didn't check whether the symbols exists in the current scope or not when we privatized loop variables.

File: flang/lib/Lower/OpenMP/Decomposer.cpp [hunk 11]
```diff
@@ -123,4 +123,9 @@ ConstructQueue buildConstructQueue(
 
   return constructs;
 }
+
+bool isLastItemInQueue(ConstructQueue::iterator item,
+                       const ConstructQueue &queue) {
+  return std::prev(queue.end()) == item;
```
Comment (kparzysz) [inline flang/lib/Lower/OpenMP/Decomposer.cpp:129]: Comparing `std::next(item) == queue.end()` is more canonical.  Operations on the `end()` iterator may not work for all containers, but the "next" comparison will.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 4 above)
Comment (kiranchandramohan) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:296]: I think we can fix this separately. Could you check whether the same issue happens with the Critical construct?

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Thanks for the refactoring here.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 4 above)
Comment (luporl) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:296]: https://github.com/llvm/llvm-project/pull/90671 fixed symbol handling in semantics, which was enough to fix https://github.com/llvm/llvm-project/issues/78936, but not https://github.com/llvm/llvm-project/issues/75767.

Maybe the issue here is similar to https://github.com/llvm/llvm-project/issues/75767.
In any case, I agree that this can be fixed separately.

**Review by luporl — APPROVED** [review]
LGTM.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 4 above)
Comment (ergawy) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:288]: Thanks both for the review.

> Could you check whether the same issue happens with the Critical construct?

That indeed causes the same issue. I will add a similar "fix" and a test before merging.

Comment (skatrak) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:288]: > I get semantic error as: `error: DEFAULT clause is not allowed on the DO directive`

I think this error happens earlier in semantics because `OMPC_Default` is not listed in the allowed clauses of `OMP_Do` in "llvm/include/llvm/Frontend/OpenMP/OMP.td". Not sure whether it should be allowed or not (making that error expected), since I'm not familiar with that part of the spec.

Comment (NimishMishra) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:288]: Thanks! Indeed it is from semantics. I was mainly checking if default privatization logic is being invoked uncharacteristically when it should not (i.e. when `default(...)` is not specified).

**Comment by omjavaid** [comment]
This has broken LLVM testsuite on various bots
https://lab.llvm.org/buildbot/#/builders/184/builds/12760
https://lab.llvm.org/buildbot/#/builders/197/builds/14376
https://lab.llvm.org/buildbot/#/builders/179/builds/10176
Its been failing for 2 days so I am going to revert this change for now. Kindly re-apply after fixing the failures.
